---
title: "Fall Into Bioinformatics"
subtitle: "Gene quantification and counts"
author: "Jimmy Breen"
date: "17/09/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    message = FALSE,
    warning = FALSE,
    fig.align = "center",
    results = "hide"
)
if (interactive()) setwd(here::here("Day_1"))
```

* TOC
{:toc}

## Session Themes and Outcomes

In today's session we're going to learn a few things for you to use when analysing RNA-seq data. For those of you that don't use RNA sequencing in your own work, this session will still be valuable to learn techniques of using counts. These could be counts of any particular annotation. For example, for people doing metagenomics, you could use counts of bacteria. If you are studying proteomics using Mass Spectromony, you can use counts of bacteria

### Summarisation of aligned data to counts
### Quality control of samples based on mapping counts
### Estimation of library sizes
### PCA/MDS assessment of outliers and batch effects

# Gene quantification

## Annotation

## GFF/GTF Files

There can be a little confusion about GFF and GTF files and these share some similarities with BED files.
GFF (General Feature Format) files have version 2 and version 3 formats, which are slightly different.
Today, we'll just look at GTF (General Transfer Format) files, which are best considered as GFF2.2, as restrictions are placed on the type of entries that can be placed in some columns.

While BED files are generally for showing all the locations of a single type of feature, multiple feature types can be specified within one of these files.
Again, like BED files, fields are tab-separated with no line provided which gives the column names.
These are fixed by design, and as such explicit column names are not required.

1. **seqname** - name of the chromosome or scaffold; chromosome names can be given with or without the 'chr' prefix. Important note: the seqname must be one used within Ensembl, _i.e._ a standard chromosome name or an Ensembl identifier such as a scaffold ID, without any additional content such as species or assembly. See the example GFF output below.
2. **source** - name of the program that generated this feature, or the data source (database or project name)
3. **feature** - feature type name, can only take the values "CDS", "start_codon", "stop_codon", "5UTR", "3UTR", "inter", "inter_CNS", "intron_CNS" and "exon" (CNS stands for Conserved Noncoding Sequence)
4. **start** - Start position of the feature, with sequence numbering starting at 1.
5. **end** - End position of the feature, with sequence numbering starting at 1.
6. **score** - A floating point value (_i.e._ decimal points are allowed)
7. **strand** - defined as + (forward) or - (reverse).
8. **frame** - One of '0', '1' or '2'. '0' indicates that the first base of the feature is the first base of a codon, '1' that the second base is the first base of a codon, and so on...
9. **attribute** - A *semicolon-separated* list of tag-value pairs, providing additional information about each feature. In the GTF format two mandatory features are required here, although they can be an empty string, _i.e._ `""`:
    + **gene_id** *value*
    + **transcript_id** *value*

Notice that there's *no real way to represent our FOXP3 sites as a GTF file*!
This format is really designed for gene-centric features as seen in the 3rd column.
An example is given below.
Also note that header rows are not controlled, but must start with the comment character `#`

```
# Data taken from http://mblab.wustl.edu/GTF22.html
381 Twinscan  CDS          380   401   .   +   0  gene_id "001"; transcript_id "001.1";
381 Twinscan  CDS          501   650   .   +   2  gene_id "001"; transcript_id "001.1";
381 Twinscan  CDS          700   707   .   +   2  gene_id "001"; transcript_id "001.1";
381 Twinscan  start_codon  380   382   .   +   0  gene_id "001"; transcript_id "001.1";
381 Twinscan  stop_codon   708   710   .   +   0  gene_id "001"; transcript_id "001.1";
```

**Note**: People variously use GFF and GTF to talk about GFF version 2, and GFF to talk about GFF version 3. GFF2 is not compatible with GFF3, so make sure you have the correct file format if you are given a GFF file. There are conversion tools available to inter-convert them, they are rarely reliable.

## Read Summarisation

Now that we have bam files with our reads aligned to the genome, the next step in a classic RNA-seq analysis would be to perform differential gene expression.
This is where we look for genes that have different expression levels between two (or more) treatment groups.
In our data, we have young (6mth) and old (24mth) samples.

Before we do this though, we need to find which reads are aligned to which gene and we will then count how many reads match each gene.
As we learned at the start of the day, gene descriptions are often contained in `gtf` files.
The full set of gene descriptions would normally be found on the [Ensembl database](ftp://ftp.ensembl.org/pub/release-95/gtf/danio_rerio/), however, given that we only have chromosome 2 today, we've placed a edited version on your VM already.
This file can be found as `~/data/Danio_rerio.GRCz11.95.chr2.gtf.gz` and now is a good time to *copy this into your directory* `~/genomes/Drerio/`.
You can use `cp` in your terminal, or use the GUI interface within RStudio for this.

### Feature Counts

There are numerous methods for counting reads which align to genes.
Common tools include HTSeq and kallisto (which is actually an aligner as well), or this can even be done inside R itself.
Today we'll use `featureCounts` from the `Subread` suite of tools as we've found it to be the fastest and most accurate tool for this task.
Here's a snapshot of the lines we use for this from one of our RNASeq pipelines.

```
## Feature Counts - obtaining all sorted bam files
SAMPLES=`find ${ALIGNDIR}/bam -name "*out.bam" | tr '\n' ' '`

## Running featureCounts on the sorted bam files
featureCounts -Q 10 \
  -s 2 \
  --fracOverlap 1 \
  -T ${CORES} \
  -a ${GTF} \
  -o ${ALIGNDIR}/counts/counts.out \ 
  ${SAMPLES}
  
## Storing the output in a single file
cut -f1,7- ${ALIGNDIR}/counts/counts.out | \
    sed 1d > ${ALIGNDIR}/counts/genes.out
```

Let's talk through these three steps:

#### 1. Find our samples
{:.no_toc}

Note that we've used a different strategy for finding our files this time.
Instead of using `$(ls ...)`, we've used `find`.
This is probably a superior approach and can withstand difficult file paths more easily.
Output from `find` will give you each result on a new line, so after this we've used `tr` to `tr`anslate line breaks `\n` to spaces.

#### 2. Count our alignments
{:.no_toc}

Here we've called `featureCounts` and passed several parameters to it.
The help page is quite extensive for this tool, so feel free to browse it in your terminal or on pages 37-42 of [the manual pdf](http://bioinf.wehi.edu.au/subread-package/SubreadUsersGuide.pdf).

| Parameter | Meaning |
|:---------:|:------- |
| `-Q 10` | The minimum mapping quality (MAPQ) score for a read to be counted. These scores can vary between aligners, but this value should capture uniquely aligning reads only. These values are generally documented very poorly & cause much confusion. |
| `-s 2`  | Specifies a reverse stranded library. Only reads which map to the opposite strand as the gene will be counted. If the library is forward stranded you would set this to `-s 1`. |
| `--fracOverlap 1` | The fraction of a read which must overlap a feature. Here every base must overlap an exon |
| `-T ${CORES}` | The number of cores we wish to use |
| `-a ${GTF}` | The gene description file |
| `-o ${ALIGNDIR}/counts/counts.out` | The output file |
| `${SAMPLES}` | The input files |

#### 3. Tidy up our output
{:.no_toc}

The output from `featureCounts` contains a few columns which we ignore, even though others may find them useful (exon coordinates, gene lengths etc)
In the final line, we're using `cut` to just return columns 1, then everything beyond 7.
`sed` is then used to delete the first line.
This will give us data in a nice easy format for importing into R.

**Try and write a script that runs `featureCounts`**.
Remember to declare all your variables, and **create any directories you need for the output**.
